### PR TITLE
Detect remnants of path.data/default.path.data bug

### LINKED ISSUE #23981: default.path.data included as a data path when path.data configured as yaml array
**Elasticsearch version**: 5.3.0

**OS version**: Tried in Centos 7 and OS X, though I suspect it doesn't matter.

**Description of the problem including expected versus actual behavior**:
When using multiple data paths, or even specifying just a single data path as an array in yaml, elasticearch includes whatever is passed to it as default.path.data, such as when installed via RPM or DEB packages, from the systemd unit OR the init.d script, since you set this when you execute the es binary:

`-Edefault.path.data=${DATA_DIR}`

**Steps to reproduce**:
Repro:

Set this in the elasticsearch.yml

```
path.data:
- "/some/datapath1"
```

run this:

`bin/elasticsearch -Edefault.path.data="/some/datapath2"`

And elasticsearch will start configured with both data paths.

**Provide logs (if relevant)**:

```
[2017-04-07T13:30:49,987][INFO ][o.e.e.NodeEnvironment    ] [jCWIeyS] using [2] data paths, mounts [[/ (/dev/disk1)]], net usable_space [198.8gb], net total_space [464.7gb], spins? [unknown], types [hfs]
```

This also happens when you use this array format:

`path.data: ["/some/datapath1"]`

### PR DESCRIPTION
**Pull request by jasontedor**
In Elasticsearch 5.3.0 a bug was introduced in the merging of default settings when the target setting existed as an array. When this bug concerns path.data and default.path.data, we ended up in a situation where the paths specified in both settings would be used to write index data. Since our packaging sets default.path.data, users that configure multiple data paths via an array and use the packaging are subject to having shards land in paths in default.path.data when that is very likely not what they intended.

This commit is an attempt to rectify this situation. If path.data and default.path.data are configured, we check for the presence of indices there. If we find any, we log messages explaining the situation and fail the node.

Closes #23981, supersedes #24052, relates #24074, relates #24093

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -724,6 +735,14 @@ public String nodeId() {
         return nodePaths;
     }
 
+    public int nodeLockId() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: can we call this `getNodeLockId()` pls

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 2]
```diff
@@ -500,6 +507,52 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
...
 
+    /**
+     * Checks for path.data and default.path.data being configured, and there being index data in any of the paths in default.path.data.
+     *
+     * @param settings the settings to check for path.data and default.path.data
+     * @param nodeEnv  the current node environment
+     * @param logger   a logger where messages regarding the detection will be logged
+     * @throws IOException if an I/O exception occurs reading the directory structure
+     */
+    static void checkForIndexDataInDefaultPathData(
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: you have 140 chars use them :)

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 3]
```diff
@@ -500,6 +507,52 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
...
+            final Settings settings,
+            final NodeEnvironment nodeEnv,
+            final Logger logger) throws IOException {
+        if (!Environment.PATH_DATA_SETTING.exists(settings) || !Environment.DEFAULT_PATH_DATA_SETTING.exists(settings)) return;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: please wrap with `{}`

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 4]
```diff
@@ -500,6 +507,52 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
...
+            for (final String availableIndexFolder : availableIndexFolders) {
+                logger.info(
+                        "index folder [{}] in default.path.data [{}] must be moved to any of {}",
+                        availableIndexFolder,
+                        nodePath.indicesPath,
+                        Arrays.stream(nodeEnv.nodePaths()).map(np -> np.indicesPath).collect(Collectors.toList()));
+            }
+        }
+
+        if (clean) return;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: please wrap {}

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 5]
```diff
@@ -500,6 +507,52 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
...
+    static void checkForIndexDataInDefaultPathData(
+            final Settings settings,
+            final NodeEnvironment nodeEnv,
+            final Logger logger) throws IOException {
+        if (!Environment.PATH_DATA_SETTING.exists(settings) || !Environment.DEFAULT_PATH_DATA_SETTING.exists(settings)) return;
+
+        boolean clean = true;
+        for (final String defaultPathData : Environment.DEFAULT_PATH_DATA_SETTING.get(settings)) {
+            final Path nodeDirectory = NodeEnvironment.resolveNodePath(getPath(defaultPathData), nodeEnv.nodeLockId());
+            if (!Files.exists(nodeDirectory)) continue;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: please wrap in {} and use `== false`

File: core/src/main/java/org/elasticsearch/node/Node.java [hunk 6]
```diff
@@ -500,6 +507,52 @@ protected Node(final Environment environment, Collection<Class<? extends Plugin>
...
+            final NodeEnvironment.NodePath nodePath = new NodeEnvironment.NodePath(nodeDirectory);
+            final Set<String> availableIndexFolders = nodeEnv.availableIndexFoldersForPath(nodePath);
+            if (availableIndexFolders.isEmpty()) continue;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/node/Node.java]: please wrap in {}

**Review by s1monw — CHANGES_REQUESTED** [review]
I am not sure this change is OK and I would like to discuss different solutions for this down the road. I'd like to not add the permission and find the possible issue earlier in the game. I think we can only do that if we ignore the nodeLockID which I think we should. It is an edgecase we are trying to solve. if we'd not look at that and only look if there is data in `default.path.data` we would do as good as this solution IMO but with much less trouble. I think this is over engineering we are trying to do here but I am ok with compromising on dropping the sophisticated method here in 6.0 since our recommendation is to upgrade to a latest version first before you upgrade even if it's a full cluster restart. We can still have a simpler check in 6.0 then. @rjernst WDYT can you review this as well?

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 5 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 6 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

File: core/src/main/java/org/elasticsearch/node/Node.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/node/Node.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: I pushed b21702e7bff6b57733908b476b87598be0ad0164.

**Review by s1monw — APPROVED** [review]

**Comment by s1monw** [comment]
@jasontedor lets merge this and I will look into improving this in master. I think the solution is good enough and we should make progress over perfection

**Pull request merged**